### PR TITLE
expression: add builtin truncate_decimal

### PR DESCRIPTION
**Pull request by niedhui**
Signed-off-by: niedhui <niedhui@gmail.com>

## What have you changed? (mandatory)

implement builtin funciton: truncate_real

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)
unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Refer to a related PR or issue link (optional)
 #3275

### CODE REVIEW

**Comment by solotzg** [comment]
Thx. LGTM.

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 1]
```diff
@@ -361,6 +361,25 @@ impl ScalarFunc {
...
+    #[inline]
+    pub fn truncate_decimal(
+        &self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<Decimal>>> {
+        let x = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let d = try_opt!(self.children[1].eval_int(ctx, row));
+        let d = if mysql::has_unsigned_flag(self.children[1].get_tp().get_flag()) {
+            (d as u64).min(127) as i8
```
Comment (solotzg) [inline src/coprocessor/dag/expr/builtin_math.rs]: Why do you choose 127 here? Generally, mysql may use 30 as the threshold. You can check on [https://www.w3schools.com/sql/trymysql.asp?filename=trysql_func_mysql_truncate](https://www.w3schools.com/sql/trymysql.asp?filename=trysql_func_mysql_truncate)

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 2]
```diff
@@ -1379,6 +1398,79 @@ mod test {
         check_truncate_data(tests);
     }
 
+    #[test]
+    fn test_truncate_decimal() {
+        let tests = vec![
```
Comment (solotzg) [inline src/coprocessor/dag/expr/builtin_math.rs]: More corner cases please.

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 1 above)
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_math.rs]: 127(i8::max_value) is the maximum i8 value,  I use 127( and -128 below) just to avoid the truncation where cast i64/u64 to i8. It is not intent to be the  max acceptable value as the threshold.
IMHO, the threshold 30 is already sit in the decimal.rs (MAX_FRACTION), and the Decimal#round will handle it (https://github.com/tikv/tikv/blob/master/src/coprocessor/codec/mysql/decimal.rs#L1154). so in here ,we just need a `valid` i8 vlaue.

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 2 above)
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_math.rs]: thanks, I found some corner case issues, will checked and fixed later

Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_math.rs]: Hi, I add thress maximum  number test cases , do you have any other corner cases to suggest, thanks?

Comment (ice1000) [reply src/coprocessor/dag/expr/builtin_math.rs]: LGTM

Comment (solotzg) [reply src/coprocessor/dag/expr/builtin_math.rs]: Thank you for committing.

Comment (solotzg) [reply src/coprocessor/dag/expr/builtin_math.rs]: Please take a look at the ci test. Use `cargo fmt` please.

Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_math.rs]: sorry, forgot to run `cargo fmt` before commit

**Comment by solotzg** [comment]
@niedhui please help to resolve conflicts. Thanks.

**Comment by niedhui** [comment]
@solotzg done

**Comment by huachaohuang** [comment]
@solotzg PTAL
